cache/instructioncache/cache.go

```diff
@@ -10,103 +11,51 @@ import (
 
 const cacheKey = "buildkit.instructioncache"
 
-type cacheGroup struct {
-	Snapshots []string `json:"snapshots"`
-}
-
 type LocalStore struct {
 	MetadataStore *metadata.Store
 	Cache         cache.Accessor
 }
 
-func (ls *LocalStore) Set(key string, refsAny []interface{}) error {
-	refs, err := toReferenceArray(refsAny)
-	if err != nil {
-		return err
-	}
-	cg := cacheGroup{}
-	for _, r := range refs {
-		cg.Snapshots = append(cg.Snapshots, r.ID())
+func (ls *LocalStore) Set(key string, value interface{}) error {
+	ref, ok := value.(cache.ImmutableRef)
+	if !ok {
+		return errors.Errorf("invalid ref")
```

Comment:
@AkihiroSuda Fixed. The problem was that as shared reference type was not correctly detected. I wonder if we should have strict types in cache interface to avoid this from happening again. I left it loose initially so that solver+cache could be tested without actual snapshot references.